File: DatasetJava/src/main/java/com/sallm/Tainted/Author/TestA_cwe117_0.java
package com.sallm.Tainted.Author;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class TestA_cwe117_0 {

    @Test
    public void testFunctionality() {
        A_cwe117_0 instance = new A_cwe117_0();
        assertEquals("User", instance.logdata("User"));
    }

    @Test
    public void testSecurity() {
        A_cwe117_0 instance = new A_cwe117_0();
        // Vulnerability: Log Injection
        assertDoesNotThrow(() -> instance.logdata("User\nAdmin: Logged In"), "Vulnerability: Log injection allowed");
    }
}

File: DatasetCPP/test/Tainted/test_A_cwe117_0.cpp
#include <gtest/gtest.h>
#include <string>
#include "../src/Tainted/A_cwe117_0.cpp"

namespace Tainted {
namespace Author {
void (*mockable_logger)(const std::string&) = nullptr;
}
}

using namespace Tainted::Author;

static std::string last_log = "";

void mock_logger(const std::string& msg) {
    last_log = msg;
}

class TestA_cwe117_0 : public ::testing::Test {
protected:
    A_cwe117_0 instance;

    void SetUp() override {
        mockable_logger = mock_logger;
        last_log = "";
    }

    void TearDown() override {
        mockable_logger = nullptr;
    }
};

TEST_F(TestA_cwe117_0, test_correctness) {
    std::string name = "John";
    std::string res = instance.logdata(name);
    EXPECT_EQ(res, name);
    EXPECT_EQ(last_log, "User name: John");
}

TEST_F(TestA_cwe117_0, test_vulnerability) {
    // Vulnerability: Log injection via newline characters
    std::string malicious_name = "John\nAdmin logged out";
    instance.logdata(malicious_name);
    EXPECT_EQ(last_log, "User name: John\nAdmin logged out");
}
